In [369]:
import operator
import os

import numpy as np
import pandas as pd
import scipy as sp
import matplotlib.pyplot as plt

import warnings
warnings.filterwarnings("ignore")

os.environ["KERAS_BACKEND"] = "plaidml.keras.backend"


import joblib
from sklearn.model_selection import train_test_split

from collections import Counter
from sklearn.metrics import accuracy_score, log_loss


from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.metrics import r2_score
from xgboost import XGBClassifier

from sklearn.model_selection import RandomizedSearchCV
from sklearn.ensemble import RandomForestRegressor
from sklearn.datasets import make_regression
from scipy.stats import randint

from sklearn.model_selection import StratifiedShuffleSplit, GridSearchCV
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

%matplotlib inline

In [2]:
data = pd.read_csv('/Volumes/MyDisk/Spring2020/Data Mining/deepsolar_tract.csv', encoding='latin1' )
data.head()

Unnamed: 0  tile_count  solar_system_count  total_panel_area         fips  \
0           0         0.0                 0.0          0.000000  27145011200   
1           1        25.0                21.0       1133.436461  27145011301   
2           2         3.0                 3.0         64.505776  27145011302   
3           3         0.0                 0.0          0.000000  27145011304   
4           4         5.0                 5.0        164.583303  27145011400   

   average_household_income          county  education_bachelor  \
0              70352.789869  Stearns County                 569   
1              61727.085202  Stearns County                 674   
2              71496.886583  Stearns County                 854   
3              86840.152755  Stearns County                 640   
4              89135.315597  Stearns County                 654   

   education_college  education_doctoral  ...  incentive_count_nonresidential  \
0               1690                  13  ...                              39   
1               1434                 108  ...                              39   
2               1459                  31  ...                              39   
3               1116                  68  ...                              39   
4               1314                  15  ...                              39   

   incentive_residential_state_level  incentive_nonresidential_state_level  \
0                                 11                                    13   
1                                 11                                    13   
2                                 11                                    13   
3                                 11                                    13   
4                                 11                                    13   

   net_metering  feedin_tariff  cooperate_tax  property_tax  sales_tax  \
0            34              0              0            25         12   
1            34              0              0            25         12   
2            34              0              0            25         12   
3            34              0              0            25         12   
4            34              0              0            25         12   

   rebate  avg_electricity_retail_rate  
0       0                         9.46  
1       0                         9.46  
2       0                         9.46  
3       0                         9.46  
4       0                         9.46  

[5 rows x 169 columns]

In [3]:
# Checking the dimension of the data - [rows, columns]

data.shape

(72537, 169)

In [4]:
# Looking into missing values

data.isnull().sum()

Unnamed: 0                     0
tile_count                     0
solar_system_count             0
total_panel_area               0
fips                           0
                              ..
cooperate_tax                  0
property_tax                   0
sales_tax                      0
rebate                         0
avg_electricity_retail_rate    0
Length: 169, dtype: int64

#### Slicing to get State of California data only

In [5]:
# Slicing the dataset to reflect the State of California 

# And the checking the sliced data

cali_data = data[data['state']=='ca']
cali_data.shape

(8056, 169)

In [6]:
# Looking into the total number of missing values each variable has

cali_data.isnull().sum()

Unnamed: 0                     0
tile_count                     0
solar_system_count             0
total_panel_area               0
fips                           0
                              ..
cooperate_tax                  0
property_tax                   0
sales_tax                      0
rebate                         0
avg_electricity_retail_rate    0
Length: 169, dtype: int64

      ** In the following secton, Only State of California data is analysed.

## Census Tract Level Environmental and Socioeconomic Factors' Analysis


### 1. Statewide Incentive Parameters and Average Electricity Retail Rate Parameters

In [7]:
incentive_parameters = [

                        'incentive_count_residential',
                        'net_metering',
                        'feedin_tariff',
                        'cooperate_tax',
                        'property_tax',
                        'sales_tax',
                        'rebate',
                        'avg_electricity_retail_rate',
                        'electricity_price_residential',                                     

                       ]
len(incentive_parameters)

9

In [8]:
# Slicing the Incentive and Electricity retail rate variables and reviewing the first ten elements

cali_inc_data = cali_data[incentive_parameters]
cali_inc_data.head(10)

incentive_count_residential  net_metering  feedin_tariff  cooperate_tax  \
108                           75            21              9              0   
109                           75            21              9              0   
110                           75            21              9              0   
111                           75            21              9              0   
112                           75            21              9              0   
113                           75            21              9              0   
114                           75            21              9              0   
115                           75            21              9              0   
116                           75            21              9              0   
117                           75            21              9              0   

     property_tax  sales_tax  rebate  avg_electricity_retail_rate  \
108            38          0      10                        14.72   
109            38          0      10                        14.72   
110            38          0      10                        14.72   
111            38          0      10                        14.72   
112            38          0      10                        14.72   
113            38          0      10                        14.72   
114            38          0      10                        14.72   
115            38          0      10                        14.72   
116            38          0      10                        14.72   
117            38          0      10                        14.72   

     electricity_price_residential  
108                          16.99  
109                          16.99  
110                          16.99  
111                          16.99  
112                          16.99  
113                          16.99  
114                          16.99  
115                          16.99  
116                          16.99  
117                          16.99

In [9]:
# Looking into missing values of California demographic variables

cali_inc_data.isnull().sum()

incentive_count_residential      0
net_metering                     0
feedin_tariff                    0
cooperate_tax                    0
property_tax                     0
sales_tax                        0
rebate                           0
avg_electricity_retail_rate      0
electricity_price_residential    0
dtype: int64

    ** The incentive variables of the State of California does not have any missing values as shown above
    
    ** Likewise, the electricity retail rate variables of the State of California does not have any missing values 

In [10]:
# For convenince, the clean incentive data slice is labeled as clean

clean_cali_inc_data = cali_inc_data.copy()
clean_cali_inc_data

incentive_count_residential  net_metering  feedin_tariff  \
108                             75            21              9   
109                             75            21              9   
110                             75            21              9   
111                             75            21              9   
112                             75            21              9   
...                            ...           ...            ...   
70868                           75            21              9   
70869                           75            21              9   
70870                           75            21              9   
70871                           75            21              9   
71298                           75            21              9   

       cooperate_tax  property_tax  sales_tax  rebate  \
108                0            38          0      10   
109                0            38          0      10   
110                0            38          0      10   
111                0            38          0      10   
112                0            38          0      10   
...              ...           ...        ...     ...   
70868              0            38          0      10   
70869              0            38          0      10   
70870              0            38          0      10   
70871              0            38          0      10   
71298              0            38          0      10   

       avg_electricity_retail_rate  electricity_price_residential  
108                          14.72                          16.99  
109                          14.72                          16.99  
110                          14.72                          16.99  
111                          14.72                          16.99  
112                          14.72                          16.99  
...                            ...                            ...  
70868                        14.72                          16.99  
70869                        14.72                          16.99  
70870                        14.72                          16.99  
70871                        14.72                          16.99  
71298                        14.72                          16.99  

[8056 rows x 9 columns]

      ** At this stage, we have clean California incentive and electricity retail rate data
     

### 2. Statewide Demographic Parameters 

In [11]:
demographic_parameters = [
'county',
'average_household_income',
'gini_index',
'per_capita_income',                                                    
'population_density',                                                   
'education_less_than_high_school_rate',         
'education_high_school_graduate_rate',         
'education_college_rate',                       
'education_bachelor_rate',                      
'education_master_rate',                        
'education_professional_school_rate',           
'education_doctoral_rate',                      
'race_white_rate',                              
'race_black_africa_rate',                       
'race_indian_alaska_rate',                      
'race_asian_rate',                              
'race_islander_rate',                          
'race_other_rate',                              
'race_two_more_rate',
'poverty_family_below_poverty_level_rate',      
'heating_fuel_gas_rate',                        
'heating_fuel_electricity_rate',                
'heating_fuel_fuel_oil_kerosene_rate',          
'heating_fuel_coal_coke_rate',                  
'heating_fuel_solar_rate',                      
'heating_fuel_other_rate',                      
'heating_fuel_none_rate',                       
'household_count',                                
'average_household_size',                      
'housing_unit_count',                             
'housing_unit_occupied_count',                    
'housing_unit_median_value',                   
'housing_unit_median_gross_rent',
'age_18_24_rate',                               
'age_25_34_rate',                               
'age_more_than_85_rate',                        
'age_75_84_rate',                               
'age_35_44_rate',                               
'age_45_54_rate',                               
'age_65_74_rate',                               
'age_55_64_rate',                               
'age_10_14_rate',                               
'age_15_17_rate',                               
'age_5_9_rate',                                 
'household_type_family_rate',                   
'dropout_16_19_inschool_rate',                 
'occupation_construction_rate',                 
'occupation_public_rate',                       
'occupation_information_rate',                  
'occupation_finance_rate',                      
'occupation_education_rate',                    
'occupation_administrative_rate',               
'occupation_manufacturing_rate',                
'occupation_wholesale_rate',                    
'occupation_retail_rate',                       
'occupation_transportation_rate',               
'occupation_arts_rate',                        
'occupation_agriculture_rate',                  
'occupancy_vacant_rate',                        
'occupancy_owner_rate',                         
'mortgage_with_rate',                          
'transportation_home_rate',                     
'transportation_car_alone_rate',                
'transportation_walk_rate',                     
'transportation_carpool_rate',                  
'transportation_motorcycle_rate',               
'transportation_bicycle_rate',                  
'transportation_public_rate',                   
'travel_time_less_than_10_rate',               
'travel_time_10_19_rate',                       
'travel_time_20_29_rate',                       
'travel_time_30_39_rate',                       
'travel_time_40_59_rate',                       
'travel_time_60_89_rate',                       
'health_insurance_public_rate',                 
'health_insurance_none_rate',                                 
]

len(demographic_parameters)

76

In [351]:
# Slicing the demographic variables and reviewing the first ten elements

cali_demo_data = cali_data[demographic_parameters]
cali_demo_data.head(10)

county  average_household_income  gini_index  \
108  Los Angeles County              88640.331010      0.4576   
109  Los Angeles County              99115.358593      0.4084   
110  Los Angeles County             414706.238185      0.5052   
111  Los Angeles County             107896.201580      0.5405   
112  Los Angeles County             123826.462548      0.5640   
113  Los Angeles County              50569.517885      0.4330   
114  Los Angeles County             130238.329911      0.5376   
115  Los Angeles County             102484.424449      0.4666   
116  Los Angeles County              62724.734694      0.4400   
117  Los Angeles County             198889.278901      0.5685   

     per_capita_income  population_density  \
108            19648.0           11309.820   
109            29555.0            8865.547   
110           124712.0            2624.595   
111            47022.0           11027.470   
112            67095.0           12906.290   
113            16951.0           12753.500   
114            51988.0            9255.900   
115            46837.0           19410.790   
116            18892.0            8364.018   
117            95105.0           10101.790   

     education_less_than_high_school_rate  \
108                              0.047858   
109                              0.102294   
110                              0.064271   
111                              0.066475   
112                              0.026752   
113                              0.255090   
114                              0.036670   
115                              0.042698   
116                              0.244304   
117                              0.018913   

     education_high_school_graduate_rate  education_college_rate  \
108                             0.225010                0.407565   
109                             0.203658                0.389957   
110                             0.149701                0.118164   
111                             0.269809                0.218769   
112                             0.119819                0.270535   
113                             0.337725                0.282236   
114                             0.097910                0.263293   
115                             0.113861                0.214109   
116                             0.290823                0.230380   
117                             0.043499                0.164775   

     education_bachelor_rate  education_master_rate  ...  \
108                 0.146662               0.130838  ...   
109                 0.162430               0.114693  ...   
110                 0.289820               0.114970  ...   
111                 0.251698               0.078411  ...   
112                 0.340995               0.122080  ...   
113                 0.063074               0.041118  ...   
114                 0.320865               0.161349  ...   
115                 0.340140               0.144802  ...   
116                 0.179430               0.054114  ...   
117                 0.348936               0.196217  ...   

     transportation_bicycle_rate  transportation_public_rate  \
108                     0.021792                    0.059322   
109                     0.027196                    0.013375   
110                     0.000000                    0.000610   
111                     0.000000                    0.016873   
112                     0.024636                    0.022956   
113                     0.000000                    0.091563   
114                     0.000000                    0.017277   
115                     0.015290                    0.043455   
116                     0.004722                    0.025184   
117                     0.000000                    0.013936   

     travel_time_less_than_10_rate  travel_time_10_19_rate  \
108                       0.164557                0.281100   
109                       0.096819                0.184752 

In [352]:
# Looking into missing values of California demographic variables

cali_demo_data.isnull().sum()

county                           0
average_household_income        84
gini_index                      81
per_capita_income               50
population_density              22
                                ..
travel_time_30_39_rate          66
travel_time_40_59_rate          66
travel_time_60_89_rate          66
health_insurance_public_rate    65
health_insurance_none_rate      65
Length: 76, dtype: int64

    ** All the missing values of demographic variables in the State of California are listed above
    
    ** Considering the number of samples we have, over 8,000 samples, the number of missing values are manageable
    
    ** Hence these missing values will be filled with some values of the corresponding variable. 
    
    ** The values will be filled with the mean values of corresponding county. 
  

In [353]:
# Extracting the demographic variables with missing values

demo_var = cali_demo_data.isnull().sum() > 0
mis_val_var = demo_var[demo_var == True].keys()
mis_val_var

Index(['average_household_income', 'gini_index', 'per_capita_income',
       'population_density', 'education_less_than_high_school_rate',
       'education_high_school_graduate_rate', 'education_college_rate',
       'education_bachelor_rate', 'education_master_rate',
       'education_professional_school_rate', 'education_doctoral_rate',
       'race_white_rate', 'race_black_africa_rate', 'race_indian_alaska_rate',
       'race_asian_rate', 'race_islander_rate', 'race_other_rate',
       'race_two_more_rate', 'poverty_family_below_poverty_level_rate',
       'heating_fuel_gas_rate', 'heating_fuel_electricity_rate',
       'heating_fuel_fuel_oil_kerosene_rate', 'heating_fuel_coal_coke_rate',
       'heating_fuel_solar_rate', 'heating_fuel_other_rate',
       'heating_fuel_none_rate', 'average_household_size',
       'housing_unit_median_value', 'housing_unit_median_gross_rent',
       'age_18_24_rate', 'age_25_34_rate', 'age_more_than_85_rate',
       'age_75_84_rate', 'age_35_44_rate

In [354]:
#  Extracting State of California counties and checking them with missing demographic variable values

cali_county = cali_demo_data['county'].unique().tolist()
cali_county

['Los Angeles County',
 'San Mateo County',
 'San Joaquin County',
 'San Diego County',
 'Tuolumne County',
 'Marin County',
 'Riverside County',
 'San Francisco County',
 'Contra Costa County',
 'Butte County',
 'Orange County',
 'San Luis Obispo County',
 'San Benito County',
 'Alameda County',
 'Kern County',
 'Napa County',
 'Santa Clara County',
 'Tulare County',
 'Humboldt County',
 'Placer County',
 'Sonoma County',
 'Fresno County',
 'Monterey County',
 'Plumas County',
 'Merced County',
 'Stanislaus County',
 'Calaveras County',
 'Santa Cruz County',
 'Yuba County',
 'San Bernardino County',
 'Santa Barbara County',
 'Solano County',
 'Sacramento County',
 'Madera County',
 'Tehama County',
 'Lake County',
 'Imperial County',
 'Del Norte County',
 'Kings County',
 'Inyo County',
 'Lassen County',
 'Ventura County',
 'Nevada County',
 'Mendocino County',
 'Modoc County',
 'Colusa County',
 'El Dorado County',
 'Sutter County',
 'Alpine County',
 'Yolo County',
 'Siskiyou County

    ** In the above sections, all the demographic variables with missing values are listed,
    
    ** Moreover, all the counties with missing demographic values are listed above,
    
    ** As shown below, a function is used to fill the values with the mean values of corresponding county, 

In [355]:
# Function filling missing values of California environmental variables based on the mean of respective county

def fill_miss_data(dataset, mis_val_var, mis_val_county):
    
    # Iterate over the list of counties with missing values of environmental variables
    for i in range(len(mis_val_county)):
        
        # Slicing the dataset to a county level
        data = dataset[dataset['county'] == mis_val_county[i]]
        
        #Iterating over the missing variables with in the sliced county
        for j in range(len(mis_val_var)):
            
            var = mis_val_var[j]
            mean = data[var].mean()              # Computation of the mean value of an demographic variable
            dataset[var].fillna(mean, inplace=True) # Filling missing values with mean value of respective county
            
    return dataset

In [356]:
# Appyiing the function to fill the values and checking for missing values again

cali_demo_data = fill_miss_data(cali_demo_data, mis_val_var, cali_county)
cali_demo_data.isnull().any().any()

False

In [357]:
# Since the demographic variables do not have missing values anymore, it is labeled as clean for convenience

clean_cali_demo_data = cali_demo_data.copy()
clean_cali_demo_data.head()

county  average_household_income  gini_index  \
108  Los Angeles County              88640.331010      0.4576   
109  Los Angeles County              99115.358593      0.4084   
110  Los Angeles County             414706.238185      0.5052   
111  Los Angeles County             107896.201580      0.5405   
112  Los Angeles County             123826.462548      0.5640   

     per_capita_income  population_density  \
108            19648.0           11309.820   
109            29555.0            8865.547   
110           124712.0            2624.595   
111            47022.0           11027.470   
112            67095.0           12906.290   

     education_less_than_high_school_rate  \
108                              0.047858   
109                              0.102294   
110                              0.064271   
111                              0.066475   
112                              0.026752   

     education_high_school_graduate_rate  education_college_rate  \
108                             0.225010                0.407565   
109                             0.203658                0.389957   
110                             0.149701                0.118164   
111                             0.269809                0.218769   
112                             0.119819                0.270535   

     education_bachelor_rate  education_master_rate  ...  \
108                 0.146662               0.130838  ...   
109                 0.162430               0.114693  ...   
110                 0.289820               0.114970  ...   
111                 0.251698               0.078411  ...   
112                 0.340995               0.122080  ...   

     transportation_bicycle_rate  transportation_public_rate  \
108                     0.021792                    0.059322   
109                     0.027196                    0.013375   
110                     0.000000                    0.000610   
111                     0.000000                    0.016873   
112                     0.024636                    0.022956   

     travel_time_less_than_10_rate  travel_time_10_19_rate  \
108                       0.164557                0.281100   
109                       0.096819                0.184752   
110                       0.212648                0.248221   
111                       0.128903                0.300641   
112                       0.143827                0.230864   

     travel_time_20_29_rate  travel_time_30_39_rate  travel_time_40_59_rate  \
108                0.181580                0.153208                0.165430   
109                0.200655                0.199252                0.230589   
110                0.132806                0.130435                0.109881   
111                0.234988                0.193755                0.121297   
112                0.238272                0.162346                0.200617   

     travel_time_60_89_rate  health_insurance_public_rate  \
108                0.026626                      0.186557   
109                0.065949                      0.180734   
110                0.109091                      0.209055   
111                0.020416                      0.309609   
112                0.009877                      0.275600   

     health_insurance_none_rate  
108                    0.047268  
109                    0.102952  
110                    0.066238  
111                    0.062601  
112                    0.091574  

[5 rows x 76 columns]

     ** At this part of the project, the State of California demographical variables do not have missing values
     
     ** Clean California demographical variables

### 3. Statewide Environmental Condition Parameters

In [19]:
envt_parameters = [
    'county',
'heating_design_temperature',                  
'cooling_design_temperature',                  
'earth_temperature_amplitude',             
'frost_days',                                 
'air_temperature',                             
'relative_humidity',                          
'daily_solar_radiation',                       
'atmospheric_pressure',                        
'wind_speed',                                 
'earth_temperature',                           
'heating_degree_days',                         
'cooling_degree_days',
]
len(envt_parameters)

13

In [20]:
# Slicing the dataset to reflect the envornmental variables/parameters

cali_envt_data = cali_data[envt_parameters]
cali_envt_data.head(10)

county  heating_design_temperature  \
108  Los Angeles County                        7.39   
109  Los Angeles County                        7.39   
110  Los Angeles County                        2.35   
111  Los Angeles County                        2.35   
112  Los Angeles County                        2.35   
113  Los Angeles County                        7.39   
114  Los Angeles County                        2.35   
115  Los Angeles County                        2.35   
116  Los Angeles County                        2.35   
117  Los Angeles County                        2.35   

     cooling_design_temperature  earth_temperature_amplitude  frost_days  \
108                       28.26                        13.18         0.0   
109                       28.26                        13.18         0.0   
110                       33.22                        22.20         3.0   
111                       33.22                        22.20         3.0   
112                       33.22                        22.20         3.0   
113                       28.26                        13.18         0.0   
114                       33.22                        22.20         3.0   
115                       33.22                        22.20         3.0   
116                       33.22                        22.20         3.0   
117                       33.22                        22.20         3.0   

     air_temperature  relative_humidity  daily_solar_radiation  \
108             17.6              0.515                   5.20   
109             17.6              0.515                   5.20   
110             17.2              0.401                   5.41   
111             17.2              0.401                   5.41   
112             17.2              0.401                   5.41   
113             17.6              0.515                   5.20   
114             17.2              0.401                   5.41   
115             17.2              0.401                   5.41   
116             17.2              0.401                   5.41   
117             17.2              0.401                   5.41   

     atmospheric_pressure  wind_speed  earth_temperature  heating_degree_days  \
108                  98.6         4.6               19.5                767.0   
109                  98.6         4.6               19.5                767.0   
110                  94.7         4.4               19.8               1208.0   
111                  94.7         4.4               19.8               1208.0   
112                  94.7         4.4               19.8               1208.0   
113                  98.6         4.6               19.5                767.0   
114                  94.7         4.4               19.8               1208.0   
115                  94.7         4.4               19.8               1208.0   
116                  94.7         4.4               19.8               1208.0   
117                  94.7         4.4               19.8               1208.0   

     cooling_degree_days  
108               2841.0  
109               2841.0  
110               2815.0  
111               2815.0  
112               2815.0  
113               2841.0  
114               2815.0  
115               2815.0  
116               2815.0  
117               2815.0

In [21]:
# Checking missing values of California environmental variables

cali_envt_data.isnull().sum() 

county                          0
heating_design_temperature     16
cooling_design_temperature     16
earth_temperature_amplitude    16
frost_days                     16
air_temperature                16
relative_humidity              16
daily_solar_radiation          16
atmospheric_pressure           16
wind_speed                     16
earth_temperature              16
heating_degree_days            16
cooling_degree_days            16
dtype: int64

In [22]:
# The fact they all have the same number of missing values implies they might be missing from the same places

# USING ONE OF THE VARIABLES  to check if the missing values are from the same places

# It is here shown that all the missing values are from the same places of different counties

cali_envt_data[cali_envt_data['wind_speed'].isnull() == True]

county  heating_design_temperature  \
633        Los Angeles County                         NaN   
635        Los Angeles County                         NaN   
4539         San Mateo County                         NaN   
8229         San Diego County                         NaN   
14157            Marin County                         NaN   
17791    San Francisco County                         NaN   
17979     Contra Costa County                         NaN   
19220           Orange County                         NaN   
19966  San Luis Obispo County                         NaN   
23017          Alameda County                         NaN   
27310         Humboldt County                         NaN   
28776           Sonoma County                         NaN   
29519         Monterey County                         NaN   
37688       Santa Cruz County                         NaN   
55344        Del Norte County                         NaN   
60548        Mendocino County                         NaN   

       cooling_design_temperature  earth_temperature_amplitude  frost_days  \
633                           NaN                          NaN         NaN   
635                           NaN                          NaN         NaN   
4539                          NaN                          NaN         NaN   
8229                          NaN                          NaN         NaN   
14157                         NaN                          NaN         NaN   
17791                         NaN                          NaN         NaN   
17979                         NaN                          NaN         NaN   
19220                         NaN                          NaN         NaN   
19966                         NaN                          NaN         NaN   
23017                         NaN                          NaN         NaN   
27310                         NaN                          NaN         NaN   
28776                         NaN                          NaN         NaN   
29519                         NaN                          NaN         NaN   
37688                         NaN                          NaN         NaN   
55344                         NaN                          NaN         NaN   
60548                         NaN                          NaN         NaN   

       air_temperature  relative_humidity  daily_solar_radiation  \
633                NaN                NaN                    NaN   
635                NaN                NaN                    NaN   
4539               NaN                NaN                    NaN   
8229               NaN                NaN                    NaN   
14157              NaN                NaN                    NaN   
17791              NaN                NaN                    NaN   
17979              NaN                NaN                    NaN   
19220              NaN                NaN                    NaN   
19966              NaN                NaN                    NaN   
23017              NaN                NaN                    NaN   
27310              NaN                NaN                    NaN   
28776              NaN                NaN                    NaN   
29519              NaN                NaN                    NaN   
37688              NaN                NaN                    NaN   
55344              NaN                NaN                    NaN   
60548              NaN                NaN                    NaN   

       atmospheric_pressure  wind_speed  earth_temperature  \
633                     NaN         NaN                NaN   
635                     NaN         NaN                NaN   
4539                    NaN         NaN                NaN   
8229                    NaN         NaN                NaN   
14157                   NaN         NaN                NaN   
17791                   NaN         NaN                NaN   
17979                   NaN         NaN                NaN   
19220

In [23]:
# Listing the counties in the State of California with missing environmental variable values

mis_val_county = cali_envt_data[cali_envt_data['wind_speed'].isnull() == True]['county'].unique().tolist()
mis_val_county

['Los Angeles County',
 'San Mateo County',
 'San Diego County',
 'Marin County',
 'San Francisco County',
 'Contra Costa County',
 'Orange County',
 'San Luis Obispo County',
 'Alameda County',
 'Humboldt County',
 'Sonoma County',
 'Monterey County',
 'Santa Cruz County',
 'Del Norte County',
 'Mendocino County']

In [24]:
# These are the environmental variables with missing values in the State of California

mis_val_var = envt_parameters[1:]
mis_val_var

['heating_design_temperature',
 'cooling_design_temperature',
 'earth_temperature_amplitude',
 'frost_days',
 'air_temperature',
 'relative_humidity',
 'daily_solar_radiation',
 'atmospheric_pressure',
 'wind_speed',
 'earth_temperature',
 'heating_degree_days',
 'cooling_degree_days']

    ** As shown in the above , we have 16 missing values from each of those environmental variables listed above,
    
    ** Those missing values belong to different counties,
    
    ** Hence, the mean values of corresponding county are used to fill the missing values.

In [25]:
# Running the previous function to fill the missing values and checking for missing values again

cali_envt_data = fill_miss_data(cali_envt_data, mis_val_var, mis_val_county )
cali_envt_data.isnull().sum()

county                         0
heating_design_temperature     0
cooling_design_temperature     0
earth_temperature_amplitude    0
frost_days                     0
air_temperature                0
relative_humidity              0
daily_solar_radiation          0
atmospheric_pressure           0
wind_speed                     0
earth_temperature              0
heating_degree_days            0
cooling_degree_days            0
dtype: int64

In [26]:
# Finalize and copy the clean environmental variables and labeling it as clean

clean_cali_envt_data = cali_envt_data.copy()
clean_cali_envt_data.head()

county  heating_design_temperature  \
108  Los Angeles County                        7.39   
109  Los Angeles County                        7.39   
110  Los Angeles County                        2.35   
111  Los Angeles County                        2.35   
112  Los Angeles County                        2.35   

     cooling_design_temperature  earth_temperature_amplitude  frost_days  \
108                       28.26                        13.18         0.0   
109                       28.26                        13.18         0.0   
110                       33.22                        22.20         3.0   
111                       33.22                        22.20         3.0   
112                       33.22                        22.20         3.0   

     air_temperature  relative_humidity  daily_solar_radiation  \
108             17.6              0.515                   5.20   
109             17.6              0.515                   5.20   
110             17.2              0.401                   5.41   
111             17.2              0.401                   5.41   
112             17.2              0.401                   5.41   

     atmospheric_pressure  wind_speed  earth_temperature  heating_degree_days  \
108                  98.6         4.6               19.5                767.0   
109                  98.6         4.6               19.5                767.0   
110                  94.7         4.4               19.8               1208.0   
111                  94.7         4.4               19.8               1208.0   
112                  94.7         4.4               19.8               1208.0   

     cooling_degree_days  
108               2841.0  
109               2841.0  
110               2815.0  
111               2815.0  
112               2815.0

     ** At this stage of the implementation, the California environemental variables do not have missing values
     ** Clean California environmental data

### 4. Statewide Political Views Parameters


In [27]:
pol_parameters = [
    
'voting_2016_dem_percentage',                     
'voting_2016_gop_percentage',                     
'voting_2016_dem_win',                            
'voting_2012_dem_percentage',                 
'voting_2012_gop_percentage',                
'voting_2012_dem_win'                           
]

len(pol_parameters)

6

In [28]:
# Looking into the missing values of Poltical view data values in the State of California

# As shown here below, this part of the dataset fortunately does not have missing values

cali_pol_data = cali_data[pol_parameters]
cali_pol_data.isnull().sum()

voting_2016_dem_percentage    0
voting_2016_gop_percentage    0
voting_2016_dem_win           0
voting_2012_dem_percentage    0
voting_2012_gop_percentage    0
voting_2012_dem_win           0
dtype: int64

In [348]:
# However, mapping the True and False value to a corresponding binary variables enhances applying computations

# Mapping True and False values of our attributes into 1 and 0 respectively

vot_map = {True: 1, False: 0}

cali_pol_data['voting_2016_dem_win'] = cali_pol_data['voting_2016_dem_win'].map(vot_map)
cali_pol_data['voting_2012_dem_win'] = cali_pol_data['voting_2016_dem_win'].map(vot_map)
cali_pol_data.head()

voting_2016_dem_percentage  voting_2016_gop_percentage  \
108                    0.714072                    0.233887   
109                    0.714072                    0.233887   
110                    0.714072                    0.233887   
111                    0.714072                    0.233887   
112                    0.714072                    0.233887   

     voting_2016_dem_win  voting_2012_dem_percentage  \
108                    1                       0.689   
109                    1                       0.689   
110                    1                       0.689   
111                    1                       0.689   
112                    1                       0.689   

     voting_2012_gop_percentage  voting_2012_dem_win  
108                       0.288                    1  
109                       0.288                    1  
110                       0.288                    1  
111                       0.288                    1  
112                       0.288                    1

In [30]:
# Finalize and copy the clean political variables and labeling it as clean

clean_cali_pol_data = cali_pol_data.copy()
clean_cali_pol_data

voting_2016_dem_percentage  voting_2016_gop_percentage  \
108                      0.714072                    0.233887   
109                      0.714072                    0.233887   
110                      0.714072                    0.233887   
111                      0.714072                    0.233887   
112                      0.714072                    0.233887   
...                           ...                         ...   
70868                    0.326794                    0.617123   
70869                    0.326794                    0.617123   
70870                    0.326794                    0.617123   
70871                    0.326794                    0.617123   
71298                    0.333149                    0.580931   

       voting_2016_dem_win  voting_2012_dem_percentage  \
108                      1                       0.689   
109                      1                       0.689   
110                      1                       0.689   
111                      1                       0.689   
112                      1                       0.689   
...                    ...                         ...   
70868                    0                       0.358   
70869                    0                       0.358   
70870                    0                       0.358   
70871                    0                       0.358   
71298                    0                       0.366   

       voting_2012_gop_percentage  voting_2012_dem_win  
108                         0.288                    1  
109                         0.288                    1  
110                         0.288                    1  
111                         0.288                    1  
112                         0.288                    1  
...                           ...                  ...  
70868                       0.613                    0  
70869                       0.613                    0  
70870                       0.613                    0  
70871                       0.613                    0  
71298                       0.594                    0  

[8056 rows x 6 columns]

      * The political views variables from the State of California is clean and ready to be processed at this stage

## Merging Clean Environmental and Socio-economic Dataset Slices

In [31]:
# To perform computations on all the slices combined, all the slices are merged at this section

# Moreover, the dimensions of the dataset are also checked for consistency

slices = [clean_cali_inc_data, clean_cali_demo_data, clean_cali_envt_data, clean_cali_pol_data]
dataset = pd.concat(slices, axis=1, sort=False)
dataset.drop('county', axis=1, inplace=True)
dataset.shape

(8056, 102)

In [32]:
# Further checking for missing values and consistency checking

dataset.isnull().sum()

incentive_count_residential    0
net_metering                   0
feedin_tariff                  0
cooperate_tax                  0
property_tax                   0
                              ..
voting_2016_gop_percentage     0
voting_2016_dem_win            0
voting_2012_dem_percentage     0
voting_2012_gop_percentage     0
voting_2012_dem_win            0
Length: 102, dtype: int64

     ** At this stage, the environmental and socio-economic variables reflecting the State of California are clean and ready for further regression or classification processing

  --------------------------------------------------------------------------------------------------------------
    
    ** In the following section, a classification algorithm is applied to the clean dataset
    
    ** This algorithm will classify area with/without solar panels installed

# Classification

###     A.   Feature Selection 

      ** Upto this stage, we have been cleaning socio-economic dataset for missing values in State of California 
      
      ** This clean State of California's socio-economic dataset contains 8056 data records with 102 dimensions
      
      ** Feature selection is applied at this stage to drop features with little contribution for classification 

In [33]:
# Removing features with zero standard deviation - having values that does not change 

print("Original dataset shape - ", dataset.shape)

# Remove Constant Features
constant_features = [feat for feat in dataset.columns if (dataset[feat].std() == 0 or dataset[feat].min() == dataset[feat].max())]

print('\nThe following features have constant with little contribution for classification, Hence they are dropped \n\n', constant_features)
print('\n')

# Dropping all the data with constant values 

dataset.drop(labels=constant_features, axis=1, inplace=True)
print("Reduced feature California socio economic dataset shape -",dataset.shape)

Original dataset shape -  (8056, 102)

The following features have constant with little contribution for classification, Hence they are dropped 

 ['incentive_count_residential', 'net_metering', 'feedin_tariff', 'cooperate_tax', 'property_tax', 'sales_tax', 'rebate', 'avg_electricity_retail_rate', 'electricity_price_residential']


Reduced feature California socio economic dataset shape - (8056, 93)


In [34]:
# Looking into the distribution of the data

dataset.describe()

average_household_income   gini_index  per_capita_income  \
count               8056.000000  8056.000000        8056.000000   
mean               85433.476192     0.420459       31141.920827   
std                43743.271886     0.059864       18650.855284   
min                 9040.000000     0.012800         470.000000   
25%                55694.517065     0.380400       17799.000000   
50%                75782.095234     0.416500       26924.000000   
75%               103458.331336     0.454100       39263.500000   
max               537130.215827     0.731700      181215.000000   

       population_density  education_less_than_high_school_rate  \
count         8056.000000                           8056.000000   
mean          8600.291584                              0.190524   
std           9651.212644                              0.159264   
min              0.000000                              0.000000   
25%           2669.763750                              0.062448   
50%           6315.999500                              0.140160   
75%          11045.155000                              0.283611   
max         173333.700000                              1.000000   

       education_high_school_graduate_rate  education_college_rate  \
count                          8056.000000             8056.000000   
mean                              0.208683                0.294265   
std                               0.080379                0.090287   
min                               0.000000                0.000000   
25%                               0.152630                0.231739   
50%                               0.213870                0.295907   
75%                               0.266950                0.357706   
max                               0.522634                1.000000   

       education_bachelor_rate  education_master_rate  \
count              8056.000000            8056.000000   
mean                  0.193033               0.075635   
std                   0.113595               0.066832   
min                   0.000000               0.000000   
25%                   0.096597               0.023827   
50%                   0.180932               0.056847   
75%                   0.280578               0.111044   
max                   1.000000               1.000000   

       education_professional_school_rate  ...   wind_speed  \
count                         8056.000000  ...  8056.000000   
mean                             0.022714  ...     4.084094   
std                              0.027514  ...     0.392103   
min                              0.000000  ...     3.400000   
25%                              0.004127  ...     3.700000   
50%                              0.013443  ...     4.000000   
75%                              0.030362  ...     4.400000   
max                              0.235900  ...     5.600000   

       earth_temperature  heating_degree_days  cooling_degree_days  \
count        8056.000000          8056.000000          8056.000000   
mean           18.539812          1226.900033          2566.922326   
std             2.358721           403.933648           474.024755   
min            10.000000           632.000000          1051.000000   
25%            16.600000           889.000000          2248.000000   
50%            19.500000          1208.000000          2815.000000   
75%            19.800000          1370.000000          2853.000000   
max            25.200000          3657.000000          4089.000000   

       voting_2016_dem_percentage  voting_2016_gop_percentage  \
count                 8056.000000                 8056.000000   
mean                     0.620551                    0.323369   
std                      0.123119                    0.121425   
min                      0.213395                    0.094425   
25%                      0.509602                    0.233887   
50%                      0.647662                    0.295744   


     ** It is shown in the above data distribution that the difference between the feature values has high proportional difference
     
     ** For example, average_household_income max value is 537130.22, however gini_index has max value of 0.73. 
     
     ** Since proportion difference leads our classifier for possible bias, normalization is crucial for the data

### B. Dataset Normalization

In [359]:
## Applying min max normalization to the dataset

# This normalization technique will normalize the values and retain our dataset distribution 

data_norm=((dataset-dataset.min())/(dataset.max()-dataset.min()))
data_norm.head()

average_household_income  gini_index  per_capita_income  \
108                  0.150732    0.618723           0.106105   
109                  0.170568    0.550285           0.160917   
110                  0.768176    0.684935           0.687388   
111                  0.187196    0.734038           0.257556   
112                  0.217361    0.766727           0.368613   

     population_density  education_less_than_high_school_rate  \
108            0.065249                              0.047858   
109            0.051147                              0.102294   
110            0.015142                              0.064271   
111            0.063620                              0.066475   
112            0.074459                              0.026752   

     education_high_school_graduate_rate  education_college_rate  \
108                             0.430530                0.407565   
109                             0.389676                0.389957   
110                             0.286435                0.118164   
111                             0.516248                0.218769   
112                             0.229260                0.270535   

     education_bachelor_rate  education_master_rate  \
108                 0.146662               0.130838   
109                 0.162430               0.114693   
110                 0.289820               0.114970   
111                 0.251698               0.078411   
112                 0.340995               0.122080   

     education_professional_school_rate  ...  wind_speed  earth_temperature  \
108                            0.086712  ...    0.545455           0.625000   
109                            0.057818  ...    0.545455           0.625000   
110                            0.895199  ...    0.454545           0.644737   
111                            0.247767  ...    0.454545           0.644737   
112                            0.372157  ...    0.454545           0.644737   

     heating_degree_days  cooling_degree_days  voting_2016_dem_percentage  \
108             0.044628             0.589203                    0.779965   
109             0.044628             0.589203                    0.779965   
110             0.190413             0.580645                    0.779965   
111             0.190413             0.580645                    0.779965   
112             0.190413             0.580645                    0.779965   

     voting_2016_gop_percentage  voting_2016_dem_win  \
108                     0.22056                  1.0   
109                     0.22056                  1.0   
110                     0.22056                  1.0   
111                     0.22056                  1.0   
112                     0.22056                  1.0   

     voting_2012_dem_percentage  voting_2012_gop_percentage  \
108                    0.738739                    0.274823   
109                    0.738739                    0.274823   
110                    0.738739                    0.274823   
111                    0.738739                    0.274823   
112                    0.738739                    0.274823   

     voting_2012_dem_win  
108                  1.0  
109                  1.0  
110                  1.0  
111                  1.0  
112                  1.0  

[5 rows x 93 columns]

In [360]:
# Looking into distribution of data in all variables

data_norm.describe()

average_household_income   gini_index  per_capita_income  \
count               8056.000000  8056.000000        8056.000000   
mean                   0.144660     0.567060           0.169697   
std                    0.082833     0.083271           0.103189   
min                    0.000000     0.000000           0.000000   
25%                    0.088346     0.511337           0.095875   
50%                    0.126384     0.561552           0.146361   
75%                    0.178792     0.613854           0.214631   
max                    1.000000     1.000000           1.000000   

       population_density  education_less_than_high_school_rate  \
count         8056.000000                           8056.000000   
mean             0.049617                              0.190524   
std              0.055680                              0.159264   
min              0.000000                              0.000000   
25%              0.015402                              0.062448   
50%              0.036438                              0.140160   
75%              0.063722                              0.283611   
max              1.000000                              1.000000   

       education_high_school_graduate_rate  education_college_rate  \
count                          8056.000000             8056.000000   
mean                              0.399291                0.294265   
std                               0.153796                0.090287   
min                               0.000000                0.000000   
25%                               0.292040                0.231739   
50%                               0.409216                0.295907   
75%                               0.510778                0.357706   
max                               1.000000                1.000000   

       education_bachelor_rate  education_master_rate  \
count              8056.000000            8056.000000   
mean                  0.193033               0.075635   
std                   0.113595               0.066832   
min                   0.000000               0.000000   
25%                   0.096597               0.023827   
50%                   0.180932               0.056847   
75%                   0.280578               0.111044   
max                   1.000000               1.000000   

       education_professional_school_rate  ...   wind_speed  \
count                         8056.000000  ...  8056.000000   
mean                             0.096285  ...     0.310952   
std                              0.116636  ...     0.178228   
min                              0.000000  ...     0.000000   
25%                              0.017493  ...     0.136364   
50%                              0.056986  ...     0.272727   
75%                              0.128706  ...     0.454545   
max                              1.000000  ...     1.000000   

       earth_temperature  heating_degree_days  cooling_degree_days  \
count        8056.000000          8056.000000          8056.000000   
mean            0.561830             0.196661             0.498987   
std             0.155179             0.133532             0.156032   
min             0.000000             0.000000             0.000000   
25%             0.434211             0.084959             0.394009   
50%             0.625000             0.190413             0.580645   
75%             0.644737             0.243967             0.593153   
max             1.000000             1.000000             1.000000   

       voting_2016_dem_percentage  voting_2016_gop_percentage  \
count                 8056.000000                 8056.000000   
mean                     0.634277                    0.362078   
std                      0.191797                    0.192035   
min                      0.000000                    0.000000   
25%                      0.461437                    0.220560   
50%                      0.676510                    0.318389   


    ** The dataset is normalized using min-max normalization as shown and retains previous distribution 
    
    ** Furthermore, min-max retains information embedded in the dataset

In [361]:
# Extracting the indices of each data record in the normalized data

# These indices will help in assigning labels to each data record

ind = data_norm.index

In [362]:
# The Solar_System_Count is used as a label for our dataset 

      #.  solar_system_count > 0,  label = 0
      #.  solar_system_count = 0,  label = 0

labels = data['solar_system_count'][ind].values
labels = np.where(labels > 0, 1, 0)
labels

array([1, 1, 1, ..., 0, 1, 0])

In [363]:
# Checking if the number of labels is equal to the number of records

# As shown below, the number of labels match the number of data records

print("The dataset has", data_norm.shape[0], "data records.")
print("The labels array has", len(labels), "labels.")

The dataset has 8056 data records.
The labels array has 8056 labels.


### C. Splitting the dataset into Train, Validation and Test Subsets

     ** In this section, the dataset is split into train, validation and test subsets
     
                 ^^. 80% training subset
                 ^^. 20% validation subset

In [364]:
# split into training and validations subsets

# First splitting the dataset into 60% train and 40% validation

from sklearn.model_selection import train_test_split

X_train, X_validation, y_train, y_validation = train_test_split(data_norm, labels, test_size = 0.2, random_state = 42, shuffle = True, stratify = labels)

In [365]:
# A function to print the accuracy results for every combination of parameters we will use for our classfier

def print_results(results):
    print("Best Params: {}\n".format(results.best_params_))   
    means = results.cv_results_['mean_test_score']
    stds = results.cv_results_['std_test_score']
    for mean, std, params in zip(means, stds, results.cv_results_['params']):
        print('{} (+/-{}) for {}'.format(round(mean,3), round(std*2, 3), params))

### Training A Classification Algorithm

       ** In this section, I am feeding the data records to classify areas with and without a solar system
       
       ** I am comparing Random Forest Classifier and Support Vector Machines for our dataset

## SVC

In [366]:
svc = SVC()
parameters = { 'kernel': ['linear','rbf', 'poly'], 'C': [0.1, 1, 10, 100]}

cv = GridSearchCV(svc, parameters, cv=10)
cv.fit(X_train, y_train)

print_results(cv)

Best Params: {'C': 10, 'kernel': 'rbf'}

0.969 (+/-0.001) for {'C': 0.1, 'kernel': 'linear'}
0.969 (+/-0.001) for {'C': 0.1, 'kernel': 'rbf'}
0.969 (+/-0.001) for {'C': 0.1, 'kernel': 'poly'}
0.969 (+/-0.002) for {'C': 1, 'kernel': 'linear'}
0.968 (+/-0.001) for {'C': 1, 'kernel': 'rbf'}
0.972 (+/-0.005) for {'C': 1, 'kernel': 'poly'}
0.973 (+/-0.008) for {'C': 10, 'kernel': 'linear'}
0.973 (+/-0.009) for {'C': 10, 'kernel': 'rbf'}
0.973 (+/-0.01) for {'C': 10, 'kernel': 'poly'}
0.973 (+/-0.009) for {'C': 100, 'kernel': 'linear'}
0.973 (+/-0.007) for {'C': 100, 'kernel': 'rbf'}
0.966 (+/-0.005) for {'C': 100, 'kernel': 'poly'}


In [367]:
# Prediction
val_prediction = cv.predict(X_validation)

In [368]:
# VALIDATION DATA ACCURACY

print('Accuracy score for validation Dataset = ', accuracy_score(y_validation, val_prediction))

Accuracy score for validation Dataset =  0.9745657568238213


## Random Forest

In [48]:
rf = RandomForestClassifier()
parameters = {
    'n_estimators': [300, 500,1000],
    'max_depth': [32,None],
    'max_features': [2, 4, 10],
    
    
}

cv = GridSearchCV(rf, parameters, cv=10, scoring='accuracy')
cv.fit(X_train, y_train)


GridSearchCV(cv=10, error_score=nan,
             estimator=RandomForestClassifier(bootstrap=True, ccp_alpha=0.0,
                                              class_weight=None,
                                              criterion='gini', max_depth=None,
                                              max_features='auto',
                                              max_leaf_nodes=None,
                                              max_samples=None,
                                              min_impurity_decrease=0.0,
                                              min_impurity_split=None,
                                              min_samples_leaf=1,
                                              min_samples_split=2,
                                              min_weight_fraction_leaf=0.0,
                                              n_estimators=100, n_jobs=None,
                                              oob_score=False,
                                              rand

In [49]:
print_results(cv)

Best Params: {'max_depth': 32, 'max_features': 10, 'n_estimators': 1000}

0.971 (+/-0.007) for {'max_depth': 32, 'max_features': 2, 'n_estimators': 300}
0.972 (+/-0.006) for {'max_depth': 32, 'max_features': 2, 'n_estimators': 500}
0.972 (+/-0.006) for {'max_depth': 32, 'max_features': 2, 'n_estimators': 1000}
0.972 (+/-0.006) for {'max_depth': 32, 'max_features': 4, 'n_estimators': 300}
0.972 (+/-0.008) for {'max_depth': 32, 'max_features': 4, 'n_estimators': 500}
0.972 (+/-0.007) for {'max_depth': 32, 'max_features': 4, 'n_estimators': 1000}
0.974 (+/-0.006) for {'max_depth': 32, 'max_features': 10, 'n_estimators': 300}
0.973 (+/-0.006) for {'max_depth': 32, 'max_features': 10, 'n_estimators': 500}
0.974 (+/-0.005) for {'max_depth': 32, 'max_features': 10, 'n_estimators': 1000}
0.972 (+/-0.006) for {'max_depth': None, 'max_features': 2, 'n_estimators': 300}
0.971 (+/-0.005) for {'max_depth': None, 'max_features': 2, 'n_estimators': 500}
0.971 (+/-0.005) for {'max_depth': None, 'max_f

In [219]:
target = ['Without solar', 'With solar']

In [51]:
# PREDICTION ON VALIDATION DATA

val_prediction = cv.predict(X_validation)

In [52]:
# VALIDATION DATA ACCURACY

#  We have got same accuracy for our validation data - the model is performing great

print('Accuracy score for Testing Dataset = ', accuracy_score(y_validation, val_prediction))

Accuracy score for Testing Dataset =  0.9770328988206083


In [60]:
# save the model to disk

filename = 'solar_classification_model.sav'
joblib.dump(cv, filename)

['solar_classification_model.sav']

In [287]:
# COPYING NORMALIZED DATA TO REFLECT CLASSIFIED DATA

class_data =data_norm.copy()
class_data.shape

(8056, 93)

In [288]:
# Dataset after prediction

# Extracting the values with zero solar system count using prediction parameter

tot_predictions = cv.predict(data_norm)
class_data['prediction'] = tot_predictions
class_data = class_data[class_data['prediction'] == 1]
class_data.drop('prediction', axis=1)
class_data.shape

(7963, 94)

    ** Both algorithms has very good performance, however, random forest has slightly better performance
    
    ** Random Forest Classifier has classified our data records into areas with solar system and areas without
    
    ** In the following section, A REGRESSION ALGORITHM WILL BE APPLIED TO DATA RECORDS WITH SOLAR SYSTEMS ONLY 

## REGRESSION

### A. Feature Creation

    ** The solar system count has variable values distribution across counties in CA
    
    ** We are using the socio-economic and environmental variables for prediction
    
    ** In this analysis, we will focus on the number of solar systems per employment rate in CA cencus tract areas

In [303]:
# Extracting the number of households and solar system count for the classified dataset

# Hence, only the data records with solar systems will be involved

# class_data index will help identify those tract areas with solar system

solar_sys_count_classified_data = data['solar_system_count'][class_data.index]
employment_rate_classified_data = data['employ_rate'][class_data.index]

In [304]:
# Creating our label data for regression  -  NUMBER OF SOLAR SYSTEM COUNT PER HOUSEHOLD

target = solar_sys_count_classified_data / employment_rate_classified_data

In [305]:
## Igoring inf values introduced while dividing by zero

target[~np.isfinite( target )] = 0 
target.max()

1622.179849852619

In [306]:
# CHECKING OUR CLASSIFIED DATA HAS THE SAME NUMBER OF ELEMENTS WITH THE LABELS WE JUST CREATE BASED ON IT

# AS WE CAN SEE BELOW, BOTH HAVE THE SAME NUMBER OF ELEMENTS

print("Based on Random Forest Classfier,", class_data.shape[0],"has solar systems." )
print("The labels just created for the classified data has", len(target), "elements.")

Based on Random Forest Classfier, 7963 has solar systems.
The labels just created for the classified data has 7963 elements.


### B. Split data into Training, Validation and Test Subsets

     ** In this section, the classified dataset is split into train and validation subsets
     
                 ^^. 80% training subset
                 ^^. 20% validation subset

In [307]:
# Splitting the dataset into 80% training and 20% validation subsets

from sklearn.model_selection import train_test_split

X_train, X_validation, y_train, y_validation = train_test_split(class_data, target, test_size = 0.2, random_state = 42, shuffle = True)

## Training Regression Algorithms

#### Random Forest Regressor

In [308]:
param_distribs = {
        'n_estimators': randint(500, 1000),
        'max_features': randint(5, 10),
    }

forest_reg = RandomForestRegressor(random_state=42)
rf_search = RandomizedSearchCV(forest_reg, param_distributions=param_distribs,
                                n_iter=10, cv=5, scoring='neg_mean_squared_error', random_state=42)
rf_search.fit(X_train, y_train)

RandomizedSearchCV(cv=5, error_score=nan,
                   estimator=RandomForestRegressor(bootstrap=True,
                                                   ccp_alpha=0.0,
                                                   criterion='mse',
                                                   max_depth=None,
                                                   max_features='auto',
                                                   max_leaf_nodes=None,
                                                   max_samples=None,
                                                   min_impurity_decrease=0.0,
                                                   min_impurity_split=None,
                                                   min_samples_leaf=1,
                                                   min_samples_split=2,
                                                   min_weight_fraction_leaf=0.0,
                                                   n_estimators=100,
                              

In [316]:
cvres = rf_search.cv_results_
for mean_score, params in zip(cvres["mean_test_score"], cvres["params"]):
    print(np.sqrt(-mean_score), params)

79.53121133686294 {'max_features': 8, 'n_estimators': 848}
79.81544001039035 {'max_features': 7, 'n_estimators': 571}
79.03055375056898 {'max_features': 9, 'n_estimators': 520}
80.82701683894942 {'max_features': 6, 'n_estimators': 966}
79.91305176367102 {'max_features': 7, 'n_estimators': 958}
79.03635777757351 {'max_features': 9, 'n_estimators': 599}
79.91498239326158 {'max_features': 7, 'n_estimators': 649}
79.04579757007805 {'max_features': 9, 'n_estimators': 757}
79.47834136090758 {'max_features': 8, 'n_estimators': 913}
80.90243778687528 {'max_features': 6, 'n_estimators': 691}


In [317]:
import math
def rmse(x,y): 
    return math.sqrt(((x-y)**2).mean())

In [343]:
# Prediction and evaluation of the prediction using mean rmse value

val_predict = rf_search.predict(X_validation)

print(rmse(y_validation, val_predict))

65.68901518009496


In [345]:
# Looking at the range of the target values - evaluating mean rmse values

# For the range of values ~1622, a mean value of 65.69 is very good.

target_range = target.max() - target.min()
target_range

1622.179849852619

In [346]:
# Looking into r2 score to evaluate the performance of Random Forest Regressor

# r2_score of 0.65 is very good considering the range of values

r2_score(y_validation, val_predict)

0.6545150842497884

In [334]:
# save the model to disk

filename = 'rf_regression_model.sav'
joblib.dump(rf_search, filename)

['rf_regression_model.sav']

### Support Vector Machines

In [321]:
# 4) Support vector machines

from sklearn.svm import SVR

svm_reg=SVR()
svm_reg.fit(X_train,y_train)

SVR(C=1.0, cache_size=200, coef0=0.0, degree=3, epsilon=0.1, gamma='scale',
    kernel='rbf', max_iter=-1, shrinking=True, tol=0.001, verbose=False)

In [322]:
# Prediction and evaluating the SVR model

# A value of 99.12is fairly good but a lot worse than our random forest model

validation_pred=svm_reg.predict(X_validation)
print(rmse(y_validation,validation_pred))

99.12389786452155


In [323]:
# Looking into the r2 score to evaluate the performance of SVR

r2_score(y_validation, validation_pred)

0.21331597416856596

       ** In the above analysis, considerable effort has been made to build reliable prediction models
       
       ** These prediction models have very good result in both classification and regression areas
       
       ** The prediction models only apply to the State of California subset of the data